File: core/transcription/src/commonMain/kotlin/dev/audiocompanion/transcription/OpenAiRealtimeProvider.kt
@file:OptIn(kotlin.io.encoding.ExperimentalEncodingApi::class)

package dev.audiocompanion.transcription

import io.ktor.client.HttpClient
import io.ktor.client.plugins.websocket.webSocket
import io.ktor.client.request.header
import io.ktor.websocket.Frame
import io.ktor.websocket.close
import io.ktor.websocket.readText
import io.ktor.websocket.send
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.channelFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlin.io.encoding.Base64

/**
 * Real-time (streaming) transcription over the OpenAI Realtime API WebSocket
 * (`wss://api.openai.com/v1/realtime?intent=transcription`). Configures a transcription session,
 * resamples the watch's 16 kHz audio to the 24 kHz the API expects, streams it as base64
 * `input_audio_buffer.append` events, and folds the `...input_audio_transcription.delta`/`.completed`
 * events into [StreamingTranscriptUpdate]s. Foreground-only; OpenAI realtime has no diarization.
 *
 * The [client] must have the ktor WebSockets plugin installed.
 */
class OpenAiRealtimeProvider(
    private val client: HttpClient,
    private val apiKey: () -> String?,
    private val cloudConsent: () -> Boolean,
    private val model: () -> String = { DEFAULT_MODEL },
    private val url: String = DEFAULT_URL,
    private val targetSampleRate: Int = TARGET_SAMPLE_RATE,
) : StreamingTranscriptionProvider {
    override val id: String = "openai-realtime"
    private val json = Json { ignoreUnknownKeys = true }

    override suspend fun isAvailable(): Boolean = cloudConsent() && !apiKey().isNullOrBlank()

    override fun transcribeStream(
        pcm: Flow<ByteArray>,
        sampleRateHz: Int,
    ): Flow<StreamingTranscriptUpdate> = channelFlow {
        if (!cloudConsent()) throw TranscriptionException.ProviderUnavailable(id)
        val key = apiKey()?.takeIf { it.isNotBlank() }
            ?: throw TranscriptionException.ProviderUnavailable(id)

        client.webSocket(
            urlString = url,
            request = {
                header("Authorization", "Bearer $key")
            },
        ) {
            send(Frame.Text(sessionConfig()))
            val sender = launch {
                pcm.collect { chunk ->
                    if (chunk.isEmpty()) return@collect
                    val resampled = PcmResampler.resampleLinearMono16(chunk, sampleRateHz, targetSampleRate)
                    send(Frame.Text(appendEvent(Base64.encode(resampled))))
                }
                send(Frame.Text("""{"type":"input_audio_buffer.commit"}"""))
            }
            val accumulator = OpenAiRealtimeAccumulator()
            try {
                for (frame in incoming) {
                    if (frame !is Frame.Text) continue
                    val event = json.decodeFromString(OpenAiRtEvent.serializer(), frame.readText())
                    when (event.type) {
                        DELTA_EVENT -> this@channelFlow.send(accumulator.delta(event.delta.orEmpty()))
                        COMPLETED_EVENT -> this@channelFlow.send(
                            accumulator.completed(event.transcript.orEmpty()),
                        )
                        ERROR_EVENT -> throw TranscriptionException.TranscriptionFailed(
                            "OpenAI realtime error: ${event.error?.message ?: "unknown"}",
                        )
                    }
                }
            } finally {
                sender.cancel()
                close()
            }
        }
    }

    private fun sessionConfig(): String =
        """
        {"type":"session.update","session":{"type":"transcription",
        "audio":{"input":{"format":{"type":"audio/pcm","rate":$targetSampleRate},
        "transcription":{"model":"${model()}"}}}}}
        """.trimIndent().replace("\n", "")

    private fun appendEvent(base64Audio: String): String =
        """{"type":"input_audio_buffer.append","audio":"$base64Audio"}"""

    @Serializable
    private data class OpenAiRtEvent(
        val type: String = "",
        val delta: String? = null,
        val transcript: String? = null,
        val error: OpenAiRtError? = null,
    )

    @Serializable
    private data class OpenAiRtError(
        val message: String? = null,
        @SerialName("type") val type: String? = null,
    )

    companion object {
        const val DEFAULT_URL = "wss://api.openai.com/v1/realtime?intent=transcription"
        const val DEFAULT_MODEL = "gpt-live-transcribe"
        const val TARGET_SAMPLE_RATE = 24_000
        private const val DELTA_EVENT = "conversation.item.input_audio_transcription.delta"
        private const val COMPLETED_EVENT = "conversation.item.input_audio_transcription.completed"
        private const val ERROR_EVENT = "error"
    }
}

/**
 * Pure folding of OpenAI realtime transcription events: `delta`s accumulate into the volatile tail;
 * a `completed` finalizes the current item into the stable transcript. Extracted for unit testing.
 */
class OpenAiRealtimeAccumulator {
    private val completed = StringBuilder()
    private val partial = StringBuilder()

    fun delta(text: String): StreamingTranscriptUpdate {
        partial.append(text)
        return update()
    }

    fun completed(transcript: String): StreamingTranscriptUpdate {
        val cleaned = transcript.trim()
        if (cleaned.isNotEmpty()) {
            if (completed.isNotEmpty()) completed.append(' ')
            completed.append(cleaned)
        }
        partial.clear()
        return update()
    }

    private fun update(): StreamingTranscriptUpdate = StreamingTranscriptUpdate(
        finalText = completed.toString().trim(),
        partialText = partial.toString().trim(),
        isFinal = false,
    )
}

File: ios/PebbleAudioKit/Sources/Transcription/Providers/OpenAiRealtimeProvider.swift
import AudioCodec
import Foundation

// Port of `core/transcription/.../OpenAiRealtimeProvider.kt`.

/// Real-time (streaming) transcription over the OpenAI Realtime API WebSocket
/// (`wss://api.openai.com/v1/realtime?intent=transcription`). Configures a transcription
/// session, resamples the watch's 16 kHz audio to the 24 kHz the API expects (linear, via
/// `PcmResampler`), streams it as base64 `input_audio_buffer.append` events, and folds the
/// `...input_audio_transcription.delta`/`.completed` events into `StreamingTranscriptUpdate`s.
/// Foreground-only; OpenAI realtime has no diarization.
public final class OpenAiRealtimeProvider: StreamingTranscriptionProvider {
    public static let defaultUrl = "wss://api.openai.com/v1/realtime?intent=transcription"
    public static let defaultModel = "gpt-live-transcribe"
    public static let targetSampleRate = 24_000

    private static let deltaEvent = "conversation.item.input_audio_transcription.delta"
    private static let completedEvent = "conversation.item.input_audio_transcription.completed"
    private static let errorEvent = "error"

    public let id = "openai-realtime"

    private let connector: any WebSocketConnector
    private let apiKey: @Sendable () -> String?
    private let cloudConsent: @Sendable () -> Bool
    private let model: @Sendable () -> String
    private let url: String
    private let targetRate: Int

    public init(
        connector: any WebSocketConnector = URLSessionWebSocketConnector(),
        apiKey: @escaping @Sendable () -> String?,
        cloudConsent: @escaping @Sendable () -> Bool,
        model: @escaping @Sendable () -> String = { OpenAiRealtimeProvider.defaultModel },
        url: String = OpenAiRealtimeProvider.defaultUrl,
        targetSampleRate: Int = OpenAiRealtimeProvider.targetSampleRate
    ) {
        self.connector = connector
        self.apiKey = apiKey
        self.cloudConsent = cloudConsent
        self.model = model
        self.url = url
        self.targetRate = targetSampleRate
    }

    public func isAvailable() async -> Bool {
        cloudConsent() && !(apiKey() ?? "").isBlank
    }

    public func transcribeStream(
        pcm: AsyncThrowingStream<Data, Error>,
        sampleRateHz: Int
    ) -> AsyncThrowingStream<StreamingTranscriptUpdate, Error> {
        AsyncThrowingStream { continuation in
            let session = Task {
                var connection: (any WebSocketConnection)?
                var sender: Task<Void, Never>?
                do {
                    guard self.cloudConsent(), let key = self.apiKey()?.nonBlank else {
                        throw TranscriptionError.providerUnavailable(providerId: self.id)
                    }
                    guard let socketUrl = URL(string: self.url) else {
                        throw TranscriptionError.transcriptionFailed("invalid OpenAI URL: \(self.url)")
                    }
                    let socket = self.connector.connect(
                        url: socketUrl, headers: ["Authorization": "Bearer \(key)"]
                    )
                    connection = socket
                    try await socket.send(text: self.sessionConfig())

                    let sourceRate = sampleRateHz
                    let targetRate = self.targetRate
                    sender = Task {
                        do {
                            for try await chunk in pcm where !chunk.isEmpty {
                                let resampled = PcmResampler.resampleLinearMono16(
                                    pcm: chunk, fromRate: sourceRate, toRate: targetRate
                                )
                                try await socket.send(
                                    text: Self.appendEvent(base64Audio: resampled.base64EncodedString())
                                )
                            }
                            try await socket.send(text: #"{"type":"input_audio_buffer.commit"}"#)
                        } catch {
                            // A send failure surfaces through the receive loop when the socket dies.
                        }
                    }

                    let decoder = JSONDecoder()
                    let accumulator = OpenAiRealtimeAccumulator()
                    while true {
                        try Task.checkCancellation()
                        guard case .text(let frame) = try await socket.receive() else { continue }
                        let event = try decoder.decode(RtEvent.self, from: Data(frame.utf8))
                        switch event.type {
                        case Self.deltaEvent:
                            continuation.yield(accumulator.delta(event.delta ?? ""))
                        case Self.completedEvent:
                            continuation.yield(accumulator.completed(event.transcript ?? ""))
                        case Self.errorEvent:
                            throw TranscriptionError.transcriptionFailed(
                                "OpenAI realtime error: \(event.error?.message ?? "unknown")"
                            )
                        default:
                            break
                        }
                    }
                } catch {
                    sender?.cancel()
                    connection?.close()
                    if error is CancellationError || error is WebSocketClosedError {
                        // Socket ended without a protocol error — the KMP flow completes normally.
                        continuation.finish()
                    } else {
                        continuation.finish(throwing: error)
                    }
                }
            }
            continuation.onTermination = { _ in session.cancel() }
        }
    }

    private func sessionConfig() -> String {
        #"{"type":"session.update","session":{"type":"transcription","#
            + #""audio":{"input":{"format":{"type":"audio/pcm","rate":\#(targetRate)},"#
            + #""transcription":{"model":"\#(model())"}}}}}"#
    }

    private static func appendEvent(base64Audio: String) -> String {
        #"{"type":"input_audio_buffer.append","audio":"\#(base64Audio)"}"#
    }

    private struct RtEvent: Decodable {
        var type: String?
        var delta: String?
        var transcript: String?
        var error: RtError?
    }

    private struct RtError: Decodable {
        var message: String?
        var type: String?
    }
}

/// Pure folding of OpenAI realtime transcription events: `delta`s accumulate into the volatile
/// tail; a `completed` finalizes the current item into the stable transcript. Extracted for unit
/// testing.
public final class OpenAiRealtimeAccumulator {
    private var completedText = ""
    private var partial = ""

    public init() {}

    public func delta(_ text: String) -> StreamingTranscriptUpdate {
        partial += text
        return update()
    }

    public func completed(_ transcript: String) -> StreamingTranscriptUpdate {
        let cleaned = transcript.trimmed
        if !cleaned.isEmpty {
            if !completedText.isEmpty { completedText += " " }
            completedText += cleaned
        }
        partial = ""
        return update()
    }

    private func update() -> StreamingTranscriptUpdate {
        StreamingTranscriptUpdate(
            finalText: completedText.trimmed,
            partialText: partial.trimmed,
            isFinal: false
        )
    }
}

extension String {
    fileprivate var trimmed: String { trimmingCharacters(in: .whitespacesAndNewlines) }
    fileprivate var isBlank: Bool { trimmed.isEmpty }
    fileprivate var nonBlank: String? { isBlank ? nil : self }
}
